test/server/listener_manager_impl_test.cc

```diff
@@ -1633,6 +1583,137 @@ TEST_F(ListenerManagerImplWithRealFiltersTest, SingleFilterChainWithSourceTypeMa
+
+  EXPECT_CALL(server_.random_, uuid());
+  EXPECT_CALL(listener_factory_, createListenSocket(_, _, _, true));
+  manager_->addOrUpdateListener(parseListenerFromV2Yaml(yaml), "", true);
+  EXPECT_EQ(1U, manager_->listeners().size());
+
+  // IPv4 client with source 10.0.1.1. No match.
+  auto filter_chain = findFilterChain(1234, "127.0.0.1", "", "tls", {}, "10.0.1.1", 111);
+  EXPECT_EQ(filter_chain, nullptr);
+
+  // IPv4 client with source 10.0.0.10, Match
+  filter_chain =
+      findFilterChain(1234, "127.0.0.1", "", "tls", {"h2", "http/1.1"}, "10.0.0.10", 111);
+  ASSERT_NE(filter_chain, nullptr);
+  EXPECT_TRUE(filter_chain->transportSocketFactory().implementsSecureTransport());
+  auto transport_socket = filter_chain->transportSocketFactory().createTransportSocket(nullptr);
+  auto ssl_socket =
+      dynamic_cast<Extensions::TransportSockets::Tls::SslSocket*>(transport_socket.get());
+  auto server_names = ssl_socket->dnsSansLocalCertificate();
+  EXPECT_EQ(server_names.size(), 1);
+  EXPECT_EQ(server_names.front(), "server1.example.com");
+
+  // IPv6 client. No match.
+  filter_chain = findFilterChain(1234, "127.0.0.1", "", "tls", {},
+                                 "2001:0db8:85a3:0000:0000:8a2e:0370:7334", 111);
+  EXPECT_EQ(filter_chain, nullptr);
+
+  // UDS client. No match.
+  filter_chain =
+      findFilterChain(0, "/tmp/test.sock", "", "tls", {"h2", "http/1.1"}, "/tmp/test.sock", 0);
+  ASSERT_EQ(filter_chain, nullptr);
+}
+
+// Verify source IPv6 matches.
+TEST_F(ListenerManagerImplWithRealFiltersTest, SingleFilterChainWithSourceIpv6Match) {
+  const std::string yaml = TestEnvironment::substitute(R"EOF(
+    address:
+      socket_address: { address: 127.0.0.1, port_value: 1234 }
+    listener_filters:
+    - name: "envoy.listener.tls_inspector"
+      config: {}
+    filter_chains:
+    - filter_chain_match:
+        source_prefix_ranges:
+          - address_prefix: 2001:0db8:85a3:0000:0000:0000:0000:0000
+            prefix_len: 64
+      tls_context:
+        common_tls_context:
+          tls_certificates:
+            - certificate_chain: { filename: "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/san_dns_cert.pem" }
+              private_key: { filename: "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/san_dns_key.pem" }
+  )EOF",
+                                                       Network::Address::IpVersion::v4);
+
+  EXPECT_CALL(server_.random_, uuid());
+  EXPECT_CALL(listener_factory_, createListenSocket(_, _, _, true));
+  manager_->addOrUpdateListener(parseListenerFromV2Yaml(yaml), "", true);
+  EXPECT_EQ(1U, manager_->listeners().size());
+
+  // IPv6 client. match.
+  auto filter_chain = findFilterChain(1234, "127.0.0.1", "", "tls", {},
+                                      "2001:0db8:85a3:0000:0000:8a2e:0370:7334", 111);
+  EXPECT_NE(filter_chain, nullptr);
+
+  // IPv6 client. No match.
```

Comment:
Nit: This comment is the same as the one above, could you note the difference? e.g. "IPv6 client with matching subnet" / "IPv6 client with non-matching subnet".